src/bus/bus.c

```diff
@@ -75,3 +76,65 @@ Peer *bus_find_peer_by_name(Bus *bus, Name **namep, const char *name_str) {
                 *namep = name;
         return peer;
 }
+
+int bus_log_policy_send(Bus *bus, uint64_t sender_id, uint64_t receiver_id, Message *message, const char *file, int line, const char *func) {
+        Log *log = bus->log;
+        int r;
+
+        log_append_common(log, LOG_WARNING, 0, file, line, func);
+
+        message_log_append(message, log);
+
+        log_appendf(log, "TRANSMIT_ACTION=SEND\n");
+
+        log_appendf(log,
+                    "SENDER_UNIQUE_NAME=:1.%llu\n"
+                    "SENDER_UID=%u\n",
+                    sender_id);
+
+        if (receiver_id == ADDRESS_ID_INVALID) {
+                log_appendf(log,
+                            "RECEIVER_UNIQUE_NAME=org.freedesktop.DBus\n");
+        } else {
+                log_appendf(log,
+                            "RECEIVER_UNIQUE_NAME=:1.%llu\n",
+                            receiver_id);
+        }
+
+        r = log_commitf(log, "Message could not be sent by :1.%llu, due to policy.", sender_id);
+        if (r)
+                return error_fold(r);
+
+        return 0;
+}
+
+int bus_log_policy_receive(Bus *bus, uint64_t receiver_id, uint64_t sender_id, Message *message, const char *file, int line, const char *func) {
+        Log *log = bus->log;
+        int r;
+
+        log_append_common(log, LOG_WARNING, 0, file, line, func);
+
+        message_log_append(message, log);
+
+        log_appendf(log, "TRANSMIT_ACTION=RECEIVE\n");
+
+        log_appendf(log,
+                    "RECEIVER_UNIQUE_NAME=:1.%llu\n"
+                    "RECEIVER_UID=%u\n",
+                    receiver_id);
+
+        if (sender_id == ADDRESS_ID_INVALID) {
+                log_appendf(log,
+                            "SENDER_UNIQUE_NAME=org.freedesktop.DBus\n");
+        } else {
+                log_appendf(log,
+                            "SENDER_UNIQUE_NAME=:1.%llu\n",
+                            sender_id);
+        }
+
+        r = log_commitf(log, "Message could not be received by :1.%llu, due to policy.", receiver_id);
+        if (r)
+                return error_fold(r);
+
+        return 0;
+}
```

Comment:
Can we make this one single function? This seems to duplicate quite a bit of code.

Reply:
Done.